app/src/org/commcare/utils/EncryptionKeyProvider.java

```diff
@@ -0,0 +1,142 @@
+            } else {
+                throw new RuntimeException("Unrecognized key type retrieved from KeyStore");
+            }
+        } else {
+            return generateKeyInKeystore(context, trueForEncrypt);
+        }
+    }
+
+    private static boolean doesKeystoreContainEncryptionKey() throws CertificateException,
+            KeyStoreException, IOException, NoSuchAlgorithmException {
+        KeyStore keystore = getKeystore();
+
+        return keystore.containsAlias(SECRET_NAME);
+    }
+
+    private static EncryptionKeyAndTransform generateKeyInKeystore(Context context, boolean trueForEncrypt)
+            throws NoSuchAlgorithmException, NoSuchProviderException, InvalidAlgorithmParameterException {
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+            KeyGenerator keyGenerator = KeyGenerator.getInstance(
+                    KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_NAME);
+            KeyGenParameterSpec keySpec = new KeyGenParameterSpec.Builder(SECRET_NAME,
+                    KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+                    .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                    .build();
+
+            keyGenerator.init(keySpec);
+            return new EncryptionKeyAndTransform(keyGenerator.generateKey(), getTransformationString(false));
+        } else {
+            KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA", KEYSTORE_NAME);
+
+            GregorianCalendar start = new GregorianCalendar();
+            GregorianCalendar end = new GregorianCalendar();
+            end.add(Calendar.YEAR, 100);
+            KeyPairGeneratorSpec keySpec = new KeyPairGeneratorSpec.Builder(context)
+                    // You'll use the alias later to retrieve the key.  It's a key for the key!
+                    .setAlias(SECRET_NAME)
+                    // The subject used for the self-signed certificate of the generated pair
+                    .setSubject(new X500Principal(String.format("CN=%s", SECRET_NAME)))
+                    // The serial number used for the self-signed certificate of the
+                    // generated pair.
+                    .setSerialNumber(BigInteger.valueOf(1337))
+                    // Date range of validity for the generated pair.
+                    .setStartDate(start.getTime())
+                    .setEndDate(end.getTime())
+                    .build();
+
+            generator.initialize(keySpec);
+            KeyPair pair = generator.generateKeyPair();
+
+            Key key = trueForEncrypt ? pair.getPublic() : pair.getPrivate();
+            return new EncryptionKeyAndTransform(key, getTransformationString(true));
+        }
+    }
+
+    @SuppressLint("InlinedApi") //Suppressing since we check the API version elsewhere
+    public static String getTransformationString(boolean useRsa) {
```

Comment:
nit: maybe pass the encryption type here directly instead if `useRsa` (`aes` vs `rsa`)

Reply:
done